migration-0.45/dashboards.go

```diff
@@ -0,0 +1,132 @@
+package main
+
+import (
+	"encoding/json"
+	"log"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	_ "github.com/mattn/go-sqlite3"
+)
+
+var (
+	db *sqlx.DB
+)
+
+type Dashboard struct {
+	Id        int       `json:"id" db:"id"`
+	Uuid      string    `json:"uuid" db:"uuid"`
+	CreatedAt time.Time `json:"created_at" db:"created_at"`
+	CreateBy  *string   `json:"created_by" db:"created_by"`
+	UpdatedAt time.Time `json:"updated_at" db:"updated_at"`
+	UpdateBy  *string   `json:"updated_by" db:"updated_by"`
+	Data      string    `json:"data" db:"data"`
+	Locked    *int      `json:"isLocked" db:"locked"`
+}
+type DashboardData struct {
+	Description string              	`json:"description"`
+	Tags        []string            	`json:"tags"`
+	Name        string    				`json:"name"`
+	Layout      []Layout            	`json:"layout"`
+	Title       string              	`json:"title"`
+	Widgets     []map[string]interface{}`json:"widgets"`
+	Variables   map[string]interface{}  `json:"variables"`
+	Version     string 				    `json:"version"`
+}
+
+type Layout struct {
+	H      int    `json:"h"`
+	I      string `json:"i"`
+	Moved  bool   `json:"moved"`
+	Static bool   `json:"static"`
+	W      int    `json:"w"`
+	X      int    `json:"x"`
+	Y      int    `json:"y"`
+}
+
+
+// initDB initalize database
+func initDB(dataSourceName string) error {
+	var err error
+
+	// open database connection
+	db, err = sqlx.Connect("sqlite3", dataSourceName)
+	return err
+}
+
+func migrateDData(data string) (string, bool) {
+	var dd *DashboardData
+	var ddNew DashboardData
```

Comment:
we can do that but it's a good practice not to modify the source of truth rather generate a new copy and run the actions on top of copy.

Reply:
> good practice not to modify the source of truth rather generate a new copy

The new copy is generated, modified, and old copy is updated to the modifed copy.
I do not think it makes a lot of sense for `source of truth` here.

If this was non-SQLite (which we eventually have plans for), we would have directly updated the DB itself.

Anyways, this is okay since this script would be running over small number of dashboards per signoz tenant.

Reply:
the new copy is generated and modified and then the new copy itself is returned right? we are not touching the first source of truth here. the point is that if we keep updating the same variable while generating the new data there could be a potential messup.